core/src/main/java/com/scalar/db/api/ConditionSetBuilder.java

```diff
@@ -0,0 +1,131 @@
+package com.scalar.db.api;
+
+import com.google.common.collect.ImmutableSet;
+import java.util.HashSet;
+import java.util.Set;
+
+public class ConditionSetBuilder {
+
+  /**
+   * Returns a builder object to build a {@code AndConditionSet} or {@code OrConditionSet}.
+   *
+   * @param condition a conditional expression to build a {@code AndConditionSet} or {@code
+   *     OrConditionSet}.
+   * @return a builder object
+   */
+  public static AndOrConditionSetBuilder condition(ConditionalExpression condition) {
+    return new AndOrConditionSetBuilder(ImmutableSet.of(condition));
+  }
+
+  /**
+   * Returns a builder object to build a {@code AndConditionSet}.
+   *
+   * @param conditions a set of conditional expressions to build a {@code AndConditionSet}.
+   * @return a builder object
+   */
+  public static BuildableAndConditionSet andConditionSet(Set<ConditionalExpression> conditions) {
+    return new BuildableAndConditionSet(conditions);
+  }
+
+  /**
+   * Returns a builder object to build a {@code OrConditionSet}.
+   *
+   * @param conditions a set of conditional expressions to build a {@code OrConditionSet}.
+   * @return a builder object
+   */
+  public static BuildableOrConditionSet orConditionSet(Set<ConditionalExpression> conditions) {
+    return new BuildableOrConditionSet(conditions);
+  }
+
+  public static class AndOrConditionSetBuilder {
+
+    private final Set<ConditionalExpression> conditions;
+
+    AndOrConditionSetBuilder(Set<ConditionalExpression> conditions) {
+      this.conditions = new HashSet<>();
+      this.conditions.addAll(conditions);
+    }
+
+    /**
+     * Adds a conditional expression for a {@code AndConditionSet}.
+     *
+     * @param condition a conditional expression for a {@code AndConditionSet}.
+     * @return a builder object
+     */
+    public BuildableAndConditionSet and(ConditionalExpression condition) {
+      conditions.add(condition);
+      return new BuildableAndConditionSet(conditions);
```

Comment:
Right, that's well-mannered. Is the following expected?

```java
  @Immutable
  public static class BuildableAndConditionSet {

    private final Set<ConditionalExpression> conditions;

    BuildableAndConditionSet(Set<ConditionalExpression> conditions) {
      this.conditions = ImmutableSet.copyOf(conditions);
    }

    public BuildableAndConditionSet and(ConditionalExpression condition) {
      return new BuildableAndConditionSet(
          new ImmutableSet.Builder<ConditionalExpression>()
              .addAll(conditions)
              .add(condition)
              .build());
    }
```

Reply:
Thanks! Unfortunately, we cannot just return `this` here since we need to prevent users from calling `or` once they start with `and`. So, recreating a set without affecting the current set seems the best solution here.

Reply:
Ah, you're right. I thought it as a usual builder pattern...

Also, I think I was thinking of defensive copy and immutability when I writing the first comment, but I forgot it when writing the second comment. Sorry for the confusion.

As we talked on another chat, immutable approach is basically better as long as there is no memory memory-pressure concerns.